vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -580,11 +582,14 @@ type internal FSharpLanguageService(package : FSharpPackage) =
             let projectFileName = site.ProjectFileName
             let projectDisplayName = projectDisplayNameOf projectFileName
 
-            let projectId = workspace.ProjectTracker.GetOrCreateProjectIdForPath(projectFileName, projectDisplayName)
+            // This projectId is not guaranteed to be the same ProjectId that will actually be created once we call CreateProjectContext
+            // in Roslyn versions once https://github.com/dotnet/roslyn/pull/26931 is merged. Roslyn will still guarantee that once
+            // there is a project in the workspace with the same path, it'll return the ID of that. So this is sufficient to use
+            // in that case as long as we only use it to call GetProject.
```

Comment:
This is horrifically icky and I feel terrible, but it avoids me having to "support" GetOrCreateProjectIdForPath for anything other than you and not tangle it on my side.

Reply:
`GetOrCreateProjectIdForPath` is used multiple times. If this is something that needs to go away, we (F#) need to do work to not rely on it entirely.

Reply:
I think that's a better way to go about it.

I'm going to try to see if I can start cleaning this stuff up now.

Reply:
Don't try _yet_ -- wait until we get the UI thread stuff merged. Then we can try moving you to the newer things directly. :smile:

Reply:
(minimally, the UI thread changes I'm including an IWorkspaceProjectContext.Id property which will remove some of the silliness here)